src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -2277,52 +2282,159 @@ void CodeGen::genBMI1Intrinsic(GenTreeHWIntrinsic* node)
             break;
         }
 
-        default:
+        case NI_BMI2_MultiplyNoFlags:
+        case NI_BMI2_X64_MultiplyNoFlags:
         {
-            unreached();
-            break;
-        }
-    }
+            int numArgs = HWIntrinsicInfo::lookupNumArgs(node);
+            assert(numArgs == 2 || numArgs == 3);
+            regNumber op1Reg = REG_NA;
+            regNumber op2Reg = REG_NA;
+            regNumber op3Reg = REG_NA;
+            regNumber lowReg = REG_NA;
+            if (numArgs == 2)
+            {
+                op1Reg = op1->gtRegNum;
+                op2Reg = op2->gtRegNum;
+                lowReg = targetReg;
+            }
+            else
+            {
+                GenTreeArgList* argList = op1->AsArgList();
+                op1                     = argList->Current();
+                genConsumeRegs(op1);
+                op1Reg  = op1->gtRegNum;
+                argList = argList->Rest();
+                op2     = argList->Current();
+                genConsumeRegs(op2);
+                op2Reg       = op2->gtRegNum;
+                argList      = argList->Rest();
+                GenTree* op3 = argList->Current();
+                genConsumeRegs(op3);
+                op3Reg = op3->gtRegNum;
+                lowReg = node->GetSingleTempReg();
+                assert(lowReg != targetReg);
+            }
+            emitAttr attr = emitTypeSize(targetType);
+            // mov the first operand into implicit source operand EDX/RDX
+            if (op1Reg != REG_EDX)
+            {
+                emit->emitIns_R_R(INS_mov, attr, REG_EDX, op1Reg);
+            }
 
-    genProduceReg(node);
-}
+            // generate code for MULX
+            if (op2->isContained() || op2->isUsedFromSpillTemp())
```

Comment:
Why can't we fixup `op2->gtRegNum` to be `EDX` and then just call into `genHWIntrinsic_R_R_RM` (given that the instruction is `reg, reg, reg/mem`)?

Reply:
Also, given that this is multiplying; I think either op1 or op2 could be in `EDX` (and the other can be contained)

Reply:
> fixup op2->gtRegNum to be EDX 

Don't you mean `op1->gtRegNum`?

The problem is the second operand register (not op2) could be an internal register (3-arg overload) or same as the target register (2-arg overload) to get the "lower half result".

Reply:
> but I still think we could do a pre-fixup and call the genHWIntrinsic_R_R_RM method

I agree that there is some duplicated code, but the better way is to make `genHWIntrinsic_R_R_RM` and similar functions more general and shareing more code with each other, rather than "pre-fixup" (or hacking) the tree nodes.

For example, I think we can refector the current containment codegen by using lambda.

Reply:
> For example, I think we can refactor the current containment codegen by using lambda.

We could also just update it to explicitly take `ins, attr, targetReg1, op1Reg, and op2` (like we did for `genHWIntrinsic_R_R_R_RM` for FMA support), which is probably significantly simpler.

Given that we basically have the following, I think the below is about as simple as we can make it (and we could revisit later and improve register allocation if possible):
* `op1` or `op2` can be contained
* The non-contained one should prefer EDX/RDX
* We have one of the two target registers, but need somewhere to put the second target (only for the three operand)

For the two operand, we can do:
```asm
emit->emitIns_R_R(INS_mov, attr, REG_EDX, nonContainedOp);
genHWIntrinsic_R_R_RM(INS_mulx, attr, targetReg, targetReg, containedOp);
```

For the three operand, we can do:
```asm
emit->emitIns_R_R(INS_mov, attr, REG_EDX, nonContainedOp);
genHWIntrinsic_R_R_RM(INS_mulx, attr, targetReg, REG_EDX, containedOp);
emit->emitIns_R_R(INS_mov, attr, op3Reg, REG_EDX);
```

For both of these cases, it requires no temporaries and only trashes `EDX/RDX`.
The initial `mov` is only required when `nonContainedOp` is not already in `EDX/RDX
The initial value of `nonContainedOp` can only be trashed if it was already in `EDX/RDX` and only for the three operand overload.

Reply:
> We could also just update it to explicitly take ins, attr, targetReg1, op1Reg, and op2

This sounds a reasonable way, will do.

> For both of these cases, it requires no temporaries and only trashes EDX/RDX.

Good point!

Reply:
> For both of these cases, it requires no temporaries and only trashes EDX/RDX.

Just a thought: We may need to assert that `targetReg != REG_EDX` for the three operand case here and if we can't guarantee that, we would need a temporary instead.

Reply:
Fixed.